monix-tail/shared/src/main/scala/monix/tail/internal/IterantZipWithIndex.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2014-2017 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import cats.effect.Sync
+import cats.syntax.all._
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
+import monix.tail.batches.BatchCursor
+import monix.tail.internal.IterantUtils._
+
+import scala.collection.mutable.ArrayBuffer
+
+private[tail] object IterantZipWithIndex {
+  /**
+    * Implementation for `Iterant#zipWithIndex`
+    */
+  def apply[F[_], A](source: Iterant[F, A])
+                    (implicit F: Sync[F]): Iterant[F, (A, Long)] = {
+
+
+    def processSeq(index: Long, ref: NextCursor[F, A]): NextCursor[F, (A, Long)] = {
+      val NextCursor(cursor, rest, stop) = ref
+      val buffer = ArrayBuffer.empty[(A, Long)]
+
+      var idx = index
+
+      while (cursor.hasNext()) {
```

Comment:
This is close to what we need, however the problem here is that a `Cursor` can be infinite and if it is, then it means that it will never get out of this while loop.

To cope with such infinite sequences described by `Cursor`, we have a value on it called `recommendedBatchSize`. So you need to do ...

```scala
var toProcess = cursor.recommendedBatchSize

while (toProcess > 0 && cursor.hasNext) {
  // ...
  toProcess -= 1
}
```

Naturally, if we do this, then we can end up with an unfinished cursor, so we then need to check for that and continue from where we've left in that case:

```scala
val next = if (cursor.hasNext) F.pure(ref) else rest
NextCursor(BatchCursor.fromAnyArray(buffer.toArray[Any]), next.map(loop(idx)), stop)
```